sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp

```diff
@@ -0,0 +1,192 @@
+//==---------------------- SPIRVLLVMTranslation.cpp ------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "SPIRVLLVMTranslation.h"
+
+#include "Kernel.h"
+#include "LLVMSPIRVLib.h"
+#include "helper/ErrorHandling.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Support/raw_ostream.h"
+#include <iostream>
+#include <sstream>
+
+using namespace jit_compiler;
+using namespace jit_compiler::translation;
+using namespace llvm;
+
+void SPIRVLLVMTranslator::getAttributeValues(std::vector<std::string> &Values,
+                                             MDNode *MD, size_t NumValues) {
+  assert(MD->getNumOperands() == NumValues && "Incorrect number of values");
+  for (const auto &MDOp : MD->operands()) {
+    auto *ConstantMD = dyn_cast<ConstantAsMetadata>(MDOp);
+    assert(ConstantMD && "Metadata not constant");
+    auto *ConstantVal = ConstantMD->getValue();
+    auto *ConstInt = dyn_cast<ConstantInt>(ConstantVal);
+    assert(ConstInt && "Not a constant integer");
+    Values.push_back(std::to_string(ConstInt->getZExtValue()));
+  }
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+static const char *REQD_WORK_GROUP_SIZE_ATTR = "reqd_work_group_size";
+// NOLINTNEXTLINE(readability-identifier-naming)
+static const char *WORK_GROUP_SIZE_HINT_ATTR = "work_group_size_hint";
+
+void SPIRVLLVMTranslator::restoreKernelAttributes(Module *Mod,
+                                                  SYCLKernelInfo &Info) {
+  auto *KernelFunction = Mod->getFunction(Info.Name);
+  assert(KernelFunction && "Kernel function not present in module");
+  if (KernelFunction->hasMetadata(REQD_WORK_GROUP_SIZE_ATTR)) {
+    auto *MD = KernelFunction->getMetadata(REQD_WORK_GROUP_SIZE_ATTR);
+    SYCLKernelAttribute ReqdAttr{REQD_WORK_GROUP_SIZE_ATTR};
+    getAttributeValues(ReqdAttr.Values, MD, 3);
```

Comment:
Please note that there is an ongoing effort to allow less than 3 operands in `reqd_work_group_size` metadata, see #7450 (@steffenlarsen can link PRs to other components)

Reply:
Thanks for pointing this out, I wasn't aware of that. We should be able to adapt our processing of the attributes once that PR is merged, to match the new behavior.

Reply:
I think for this particular case https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1726 is also relevant. Note that in the reverse-translator case you will always get 3 operands back, but I believe you also care about generated LLVM IR that hasn't gone through the translator, in which case you may need to adjust for optional parameters.

Reply:
Thanks for the hint!

Right now, we only care about consuming SPIR-V. Untranslated LLVM IR will become relevant for targets that the DPC++ frontend currently does not compile to SPIR-V, e.g., targeting CUDA or AMD GPUs. Will keep this in mind for when we work on that.